ci-operator/step-registry/rosa/nodes/wait-ready/rosa-nodes-wait-ready-commands.sh

```diff
@@ -10,24 +10,109 @@ log(){
+
+# Display details of machinesets and nodes for Classic Rosa cluster
+function listMachineAndNodeDetails() {
+    echo "List machine details"
+    # Get current machinesets and machines 
     log "$(date) - List of machinesets"
-    oc get machinesets -A
+    echo "oc get machinesets -n openshift-machine-api"
+    oc get machinesets -n openshift-machine-api
     log "$(date) - List of machines"
-    oc get machines -A | cat -n
-    log "$(date) - List of all worker only nodes, excluding infra"
-    oc get nodes --no-headers -l node-role.kubernetes.io/worker,node-role.kubernetes.io/infra!= | cat -n
+    echo "oc get machines -n openshift-machine-api | cat -n"
+    oc get machines -n openshift-machine-api | cat -n
 
-    # Get details of machines not in Running state
+    # Get details of worker machines not in Running state
+    log "$(date) - Worker machines not in Running state, if any"
     for machine in $(oc get machine  -n openshift-machine-api -l machine.openshift.io/cluster-api-machine-type=worker,machine.openshift.io/cluster-api-machine-type!=infra  --output jsonpath="{.items[?(@.status.phase!='Running')].metadata.name}"); do
       oc describe machine "$machine" -n openshift-machine-api
     done
-    # Get details of nodes not in Ready state
-    for node in $(oc get nodes --no-headers -l node-role.kubernetes.io/worker,node-role.kubernetes.io/infra!= --output jsonpath="{.items[?(@.status.conditions[-1].type!='Ready')].metadata.name}"); do
-      oc describe node "$node"
+    echo
+    
+    # Get details of all nodes 
+    listNodeDetails
+}
+
+# Display Machine Config Pool details
+function listMachineConfigPoolDetails() {
+    log "$(date) - List machine config pools"
+    echo "oc get mcp"
+    oc get mcp
+}
+
+# List details of machinesets, machines and nodes depending on Classic Rosa or HCP cluster
+function listDetails() {
+  if [ "$is_hcp_cluster" = "false" ]; then
+    echo "Listing machine pool config, machine and node details"
+    listMachineConfigPoolDetails
+    listMachineAndNodeDetails
+  else
+    echo "Listing node details only"
+    listNodeDetails
+  fi
+}
+
+# Function introducing sleep and wait for worker nodes to become Ready state
+function waitForReady() {
+    echo "Wait for all nodes to be Ready"
+
+    # Sleep for 2 minutes to allow for the nodes to begin restarting after initial scaling
+    sleep 120
+    # Query the node state until all of the nodes are ready
+    FINAL_NODE_STATE="Pass"
+    max_attempts=60 # Max of 60 attempts with 30 sec wait time in between, for total of 30 min
+    for i in $( seq $max_attempts ); do
+        NODE_STATE="$(oc get nodes || echo "ERROR")"
+        if [[ ${NODE_STATE} == *"NotReady"*  || ${NODE_STATE} == *"SchedulingDisabled"* ]]; then
+            FINAL_NODE_STATE="Fail"
+            echo "Not all nodes have finished restarting - waiting for 30 seconds, attempt ${i}"
+            sleep 30
+        elif [[ ${NODE_STATE} == "ERROR" ]]; then
+            retry_api_count=0
```

Comment:
There is a logic error, you should init `retry` at the top of the loop, otherwise, retry will never reach its limitation.

Reply:
@yasun1 If we init retry at the top of the loop (i.e., above `for i in $( seq $max_attempts )`), it means that throughout the 30 min, we only have 3 chances for API calls to fail, whether those 3 chances are consecutive attempts to retry API or not.

In my logic here, I am saying that when we hit an API error, we try 3 times consecutively and if we continue to get API error after 3 attempts, then exit the program.

Reply:
In general, validating all this is complex, but I feel that when we hit an API error, we should try 3 times consecutively before giving up and exiting the test. At max we wait 1 min, 30 sec (3 consecutive attempts) before giving up the test and cluster.

Reply:
In your logic, you do not have the chance to retry 3 times. Every time, you reset retry_api_count to 0, the value is changed between 0 and 1. 

For the error of the command `oc get node`,  we add retry to avoid some accidental scenarios(etc. network issue), if retry 3 times(you can set higher to retry more times), if still meets the error, there should be something wrong on the cluster, we should exit but not still try in 30 minutes.